src/base/baseBrushTool.js

```diff
@@ -0,0 +1,367 @@
+import external from './../externalModules.js';
+import EVENTS from './../events.js';
+import baseTool from './../base/baseTool.js';
+// Utils
+import KeyboardController from '../fancy-tools/shared/KeyboardController.js';
+import isToolActive from '../fancy-tools/shared/isToolActive.js';
+import { getNewContext } from '../util/drawing.js';
+import { COLOR_MAP_ID } from '../stateManagement/brushToolColors.js';
+
+export default class extends baseTool {
+  constructor ({
+    name,
+    strategies,
+    defaultStrategy,
+    configuration,
+    supportedInteractionTypes
+  }) {
+    super({
+      name,
+      strategies,
+      defaultStrategy,
+      configuration,
+      supportedInteractionTypes
+    });
+
+    this.isBrushTool = true;
+    this.hasCursor = true;
+    this._referencedToolData = 'brush';
+
+    this._drawingMouseUpCallback = this._drawingMouseUpCallback.bind(this);
+  }
+
+  //===================================================================
+  // Abstract Methods - Must be implemented.
+  //===================================================================
+
+  /**
+  * Helper function for rendering the brush.
+  *
+  * @abstract
+  * @param {Object} evt - The event.
+  */
+  renderBrush (evt) {
+    throw new Error(`Method renderBrush not implemented for ${this.toolName}.`);
+  }
+
+  /**
+   * Paints the data to the canvas.
+   *
+   * @protected
+   * @abstract
+   * @param  {Object} eventData The data object associated with the event.
+   */
+  _paint (eventData) {
+    throw new Error(`Method renderBrush not implemented for ${this.toolName}.`);
+  }
+
+  //===================================================================
+  // Virtual Methods - Have default behavior but may be overriden.
+  //===================================================================
+
+  /**
+   * Callback for when the tool is activated.
+   *
+   * @virtual
+   */
+  activeCallback () {
+    const configuration = this.configuration;
+
+    this._changeDrawColor(configuration.draw);
+  }
+
+  /**
+  * Event handler for MOUSE_DRAG event.
+  *
+  * @virtual
+  * @event
+  * @param {Object} evt - The event.
+  */
+  mouseDragCallback (evt) {
+    this._startPainting(evt);
+  }
+
+  /**
+  * Event handler for MOUSE_DOWN event.
+  *
+  * @virtual
+  * @event
+  * @param {Object} evt - The event.
+  */
+  postMouseDownCallback (evt) {
```

Comment:
Interesting choice. I wondered if you'd favor `brush` or `annotations` when trying to paint near/over an existing annotation.

Reply:
Yeah I'm still unsure of this.The two options are:
- Completely ignore the passive status of all other tools.
- Make painting potentially more fiddly on a densely annotated image.

This is something we should probably discuss on Thursday.

Reply:
Feel free to create a discussion item! 🎉